ares_library_init_android.3

```diff
@@ -0,0 +1,152 @@
+to use the Connectivty Manager registered using
+\fIares_library_android_init\fP. This must be call before
+\fIares_library_android_init\fP.
+.PP
+The
+.B ares_library_android_initialized
+function can be used to check whether c-ares has been initalized for use
+with Android.
+.SH RETURN VALUES
+ARES_SUCCESS will be returned on success otherwise an error code will
+be returned.
+.SH THREAD SAFETY
+.B These init functions are not thread safe.
+You have to call it once the program has started, but this call must be done
+before the program starts any other thread. This is required to avoid
+potential race conditions in library initialization, and also due to the fact
+these might call functions from other libraries that
+are thread unsafe, and could conflict with any other thread that is already
+using these other libraries.
+.SH JNI
+Accesing the Connectivity Manager though Java:
+.PP
+Register the \fIares_library_android_init\fP.
+.PP
+.Bd -literal
+  static JNINativeMethod funcs[] = {
+  { "initialize_native",     "(Landroid/net/ConnectivityManager;)I",
+    (void *)&ares_library_android_init}
+  };
+
+  JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *reserved)
+  {
+    JNIEnv *env = NULL;
+    jclass  cls = NULL;
+    jint    res;
+  
+    if ((*vm)->GetEnv(vm, (void **)&env, JNI_VERSION_1_6) != JNI_OK)
+      return -1;
+  
+    cls = (*env)->FindClass(env, JNIT_CLASS);
+    if (cls == NULL)
+      return -1;
+  
+    res = (*env)->RegisterNatives(env, cls, funcs, sizeof(funcs)/sizeof(funcs[0]));
+    if (res != 0)
+      return -1;
+  
+    ares_library_init_jvm(vm);
+    return JNI_VERSION_1_6;
+  }
+.Ed
+.PP
+Calling the registered function from Java:
+.PP
+.Bd -literal
+  public class MyObject {
+    static {
+      System.loadLibrary("cares");
+    }
+  
+    private static native boolean initialize_native(ConnectivityManager
+      connectivity_manager);
+  
+    public static boolean initialize(Context context) {
+      initialize_native((ConnectivityManager)context.getSystemService(Context.CONNECTIVITY_SERVICE));
+    }
+  }
+.Ed
+.PP
+Initalizing the Connectivity Manager in JNI directly using an Android Context. It is assumed
+the JVM has aleady been registered through \fIJNI_OnLoad\fP.
+.PP
+.Bd -literal
+  void initialize(jobject android_context)
+  {
+    jclass obj_cls = jni_get_class(env, "android/content/Context");
+    jmethodID obj_mid = jni_get_method_id(env, obj_cls, "getSystemService", "(Ljava/lang/String;)Ljava/lang/Object;");
+    jfieldID fid = (*env)->GetStaticFieldID(env, obj_cls, "CONNECTIVITY_SERVICE", "Ljava/lang/String;");
+    jstring str = (*env)->GetStaticObjectField(env, obj_cls, fid);
+    connectivity_manager = (*env)->CallObjectMethod(env, android_context, obj_mid, str);
+    if (connectivity_manager == NULL)
+      return;
+    ares_library_android_init(connectivity_manager);
+  }
+.Ed
+.SH AVAILABILITY
+This function was first introduced in c-ares version 1.?.?.
```

Comment:
I suspect we'll forget to come back and update this, so maybe fill it in as 1.14.0 as a best guess?

Reply:
Yes, please make a best guess to cover for us missing to update this later... =)